### PR TITLE
Recalculate active quotes which have disabled rule ID among applied r…

### PR DESCRIPTION
**Pull request by zhartaunik**
…24526)

### Description (*)

I used event "save_commit_after" to do not prevent Rule object saving if something will be wrong with quote saving (customer may have too many quotes with applied described rule)

### Fixed Issues (if relevant)

1. magento/magento2#24526: quote_item.applied_rule_ids not updated after disabling cart price rule

### Manual testing scenarios (*)

1. Create new rule
2. Add product to cart
3. Disable rule
4. Reload cart

### Contribution checklist (*)
 - [*] Pull request has a meaningful description of its purpose
 - [*] All commits are accompanied by meaningful commit messages
 - [*] All new or changed code is covered with unit/integration tests (if applicable)
 - [*] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by zhartaunik** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @zhartaunik. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @zhartaunik, here is your [new Magento instance](https://pr-24921.instances.magento-community.engineering).
Admin access: [https://pr-24921.instances.magento-community.engineering/admin](https://pr-24921.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by zhartaunik** [comment]
Double-checked on your test environment. Fix works well

File: app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php
Comment (rogyar) [inline app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php]: Hi. I will just add my 5 cents here. That's an interesting implementation. However, if we imagine a case with a large result set of quotes with the applied rule ID, this iterations loop might be really heavy for the system since collecting totals and saving the quote are really heavyweight operations. If we need to perform this action for 100-200 active quotes, we might end up with gateway timeout (best scenario).
Ideally, such actions should be delegated to the message queue and processed asynchronously. 
Just thinking out loud.

Comment (zhartaunik) [reply app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php]: Make sense. I thought that there is no rabbitMQ in "open source" edition... Is it good practice at all, as many users may not use ampq?

Comment (paliarush) [reply app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php]: I would suggest skipping totals calculation here at all. Just update quote to remove outdated rule.

Totals collection will be triggered on storefront when the user reloads his cart or performs any action on it.

After we verify that this works as expected, we may put message to queue to defer this action.

Comment (fooman) [reply app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php]: There is the field trigger_recollect which I believe we could use for this.

Comment (zhartaunik) [reply app/code/Magento/SalesRule/Observer/UpdateQuoteOnDisableSalesrule.php]: PR updated, tested with the condition ```'entity_id IN(?)' => range(0, 1000000)```

**Comment by zhartaunik** [comment]
Can anybody review it?? Please advice am I doing something wrong? Should I flag my PR somehow?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @zhartaunik,
Your changes looks good. Could you cover your changes with MFTF test?

**Comment by zhartaunik** [comment]
> Hi @zhartaunik,
> Your changes looks good. Could you cover your changes with MFTF test?

Sure! Will do

**Comment by zhartaunik** [comment]
Hi @ihor-sviziev 
I have created a test and checked it's work. With a disabled fix, the test fails as expected. With applied fix the test passing successfully. But anyway I need some advice:

- I didn't know which testCaseId to add and left this field empty: &lt;testCaseId value=""/&gt;
- The created test had several repeatable blocks and most likely they should be moved to ActionGroups, but other sibling tests should also use them. According to the contributing policy, I should not modify tests that aren't related to the current task. And new ActionGroups creation isn't justified.
- With an added positive test maybe it makes sense to add a negative test: When we disable a rule which isn't related to the current quote, 'trigger_recollect' flag shouldn't be added. But I'm not sure about the best way to test. No progress bar near the totals section?

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6155](https://jira.corp.magento.com/browse/ENGCOM-6155) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@zhartaunik as for me this test that you added it good enough, not sure we need to check some negative scenario

**Comment by engcom-Bravo** [comment]
Hello @zhartaunik. Thank You for Your contribution. Could You please a bit improve Your PR in the following way. You see it works well with the here described scenario, i.e. 
1. Create new rule
2. Add product to cart
3. `Disable`rule
4. Reload cart.

But if we slightly change the scenario in step 3, the PR currently won't work, though logically is also supposed to. So if we put it this way ('Disable' is swapped to 'Delete' in step 3):

1. Create new rule
2. Add product to cart
3. `Delete` rule
4. Reload cart.

**Expected Result**
quote_item.applied_rule_ids is NULL.

**Actual Result**
quote_item.applied_rule_ids contains deleted rule ID.

I thing the PR should better cover case with "Delete" too. Correct me if I'm wrong. Thank You.

**Comment by zhartaunik** [comment]
> Hello @zhartaunik. Thank You for Your contribution. Could You please a bit improve Your PR in the following way...

Will do

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDeleteAppliedRuleRecalculateQuote.xml
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDeleteAppliedRuleRecalculateQuote.xml]: This is actually deletes the rule. please update comment.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDisableAppliedRuleRecalculateQuote.xml
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDisableAppliedRuleRecalculateQuote.xml]:
```suggestion
            <title value="Quote Totals should be invalidated if cart rule was applied to it and then was disabled." />
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Bravo** [comment]
@zhartaunik Thank You for the work You've done, but there's another crucial improvement begs for itself to be implemented and I'm really sorry to not have thought about it before I asked You to improve the PR last time. So lets consider such steps to reproduce:

1. Add product to cart. Take a note at Order Total and remember it.
2. Create new rule or activate an existing one.
3. Reload cart. Watch Order Total and notice - the rule has been applied.
4. Watch quote_item.applied_rule_ids.

**Expected Result**
It equals the rule ID.

**Actual Result**
It equals NULL.

Could You please do that too? Thank You in advance.

**Comment by zhartaunik** [comment]
Will do soon

**Comment by zhartaunik** [comment]
When I was thinking about the fix I had a look at the workflow of "Update Shopping Cart" functionality. It calls ```\Magento\Checkout\Model\Cart::save()```
There after collectTotals() method executes ```$this->quoteRepository->save($this->getQuote());```

The described above scenario was double-checked.

1. Create new simple product and two new cart rules which should trigger on this product anytime (I hang them on subtotal > 5)
2. Open cart page. Both rules were applied. _**(Database: applied_rule_ids: 1,2)**_ Quote is always recollecting here and now it's also saving.
3. Disabled 2nd rule in admin area _**(Database: applied_rule_ids: 1,2)**_. Two rules are still applied.
4. Open cart page. Only the first rule was applied _**(Database: applied_rule_ids: 1)**_ Currently quote recollecting totals every time, but it wasn't saved. Save added and 2nd rule removed.
5. Enable 2nd rule again via the admin area. _**(Database: applied_rule_ids: 1)**_ Nothing happens yet until we open the shopping cart.
6. Open cart page. _**(Database: applied_rule_ids: 1,2)**_ Totals are recalculated as usual on this page and the quote is now saving.

File: app/code/Magento/Checkout/Model/TotalsInformationManagement.php
Comment (zhartaunik) [inline app/code/Magento/Checkout/Model/TotalsInformationManagement.php:74]: When I was thinking about the fix I had a look at the workflow of "Update Shopping Cart" functionality. It calls ```\Magento\Checkout\Model\Cart::save()```
There after collectTotals() method executes ```$this->quoteRepository->save($this->getQuote());```

**Comment by zhartaunik** [comment]
Hi @ihor-sviziev 
I totally changed my fix, please have a look. What do you think?

**Comment by ihor-sviziev** [comment]
I see that some functional tests became failing, need to review them. Unfortunately tests results isn't available now, so I can't see what exactly failed. Let's wait until it will be fixed

**Comment by zhartaunik** [comment]
Hi @ihor-sviziev 
It looks like test results are resurrected. 
Regarding static tests there is nothing referred to my fix. However if it needed I can apply necessary changes. What should I do in such cases? Is there any guide which can clarify such edge cases?

Functional test results look strange. When I have possibility I will doublecheck them. Though result looks unbelievably.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @zhartaunik,
Please review failing functional tests, even after restart they became broken (even in of the tests magento report was thrown).

Could you check it?

Also it's not clear for me why did you decided to fully change your fix? Looks like current implementation could cause infinite recursive call of collect totals.

**Comment by zhartaunik** [comment]
Hi @ihor-sviziev 
Actually it looks like totals calculation was fixed in another PR but appeared another issue, which is described in the following message: https://github.com/magento/magento2/pull/24921#issuecomment-546409224

If you check vanilla Magento you should see that after enabling and disabling the rule, the totals section works correctly. But there is inconsistency in the database.

Initially, there were two issues:
1. Totals recalculation was triggered in JS only in specific cases (very old code):
> Magento/Checkout/view/frontend/web/js/model/cart/totals-processor/default.js
```
            if (!cartCache.isChanged('cartVersion', customerData.get('cart')()['data_id']) &&
                !cartCache.isChanged('shippingMethodCode', data.shippingMethodCode) &&
                !cartCache.isChanged('shippingCarrierCode', data.shippingCarrierCode) &&
                !cartCache.isChanged('address', address) &&
                cartCache.get('totals')
            ) {
```
This is replaced now with the code which trigger totals recollect every time
2. Totals were recalculated incorrecly. This most likely was fixed here: 
> \Magento\SalesRule\Model\Quote\Discount
> Prabhu Ram 17.09.2019, 19:42 MC-19254: Implement the schema changes - Review fixes for Bic/ Added new resolver for cart discounts 4d160b18
> Prabhu Ram 13.09.2019, 19:37 MC-19254: Implement the schema changes - review fixes 26630b9e
> Prabhu Ram 12.09.2019, 23:57 MC-19254: Implement the schema changes - test fixes a78b6355
> Prabhu Ram 10.09.2019, 16:08 MC-19254: Implement the schema changes - Added discount breakdown at cart level 671ae981

**And now UI represents the actual result.**

So the fix should just save the quote after recollect. The row with collecting can be removed because quote saving triggers recollecting by default.

I'm not ready to answer what is going on with tests, need 1 more hour to get their names and run locally.

**Comment by zhartaunik** [comment]
Hi @ihor-sviziev 
Cc @vital-pantsialeyeu 
Sorry, I have no enough time to proceed with the fix. Here is what I found why the test is falling down:

Fatal error: Uncaught Error: Call to a member function getGiftMessageId() on null in /var/www/magento/app/code/Magento/GiftMessage/Block/Message/Multishipping/Plugin/ItemsBox.php:51 Stack trace: #0 /var/www/magento/lib/internal/Magento/Framework/Interception/Interceptor.php(152): Magento\GiftMessage\Block\Message\Multishipping\Plugin\ItemsBox->afterGetItemsBoxTextAfter(Object(Magento\Multishipping\Block\Checkout\Shipping\Interceptor), '', Object(Magento\Quote\Model\Quote\Address)) #1 /var/www/magento/lib/internal/Magento/Framework/Interception/Interceptor.php(159): Magento\Multishipping\Block\Checkout\Shipping\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Quote\Model\Quote\Address)) #2 /var/www/magento/generated/code/Magento/Multishipping/Block/Checkout/Shipping/Interceptor.php(182): Magento\Multishipping\Block\Checkout\Shipping\Interceptor->___callPlugins('getItemsBoxText...', Array, Array) #3 /var/www/magento/app/code/Magento/Multishipping/view/frontend/templates/checkout/shipping.phtml(86): Magen in /var/www/magento/app/code/Magento/GiftMessage/Block/Message/Multishipping/Plugin/ItemsBox.php on line 51

Here is revision: 8c2760f4b09ff63188c25ae131265c5dae60075b (Fresh fix, done month ago).

File: app/code/Magento/Checkout/Model/TotalsInformationManagement.php
Comment (ihor-sviziev) [reply app/code/Magento/Checkout/Model/TotalsInformationManagement.php:74]: Hi @paliarush,

I'm not sure if it's ok to have cart totals save instead of just collect totals

Could you review and confirm this behavior change?

**Comment by ihor-sviziev** [comment]
Waiting for approval/feedback from architect, for now I put this PR on hold: https://github.com/magento/magento2/pull/24921/files#r342952196

**Comment by zhartaunik** [comment]
When I was thinking about this solution - I find that saving quote is idempotent. In other words, any save quote operation should not break the overall situation.

Why I had to add an extra condition for multi shipping? The steps to reproduce are the following:
The customer goes to multi checkout, saves shipping information and then decided to return back to cart in another window. Without added condition (in my code), Magento adds an additional shipping address. And to already existing 2 shipping addresses and 1 billing address (we added them on multi-checkout) we adding one more shipping and billing address. And as a result, the quote has 5 addresses, which is absolutely wrong.

But let's remind what issue are we fining here? Left only this one: 
https://github.com/magento/magento2/pull/24921#issuecomment-546409224
And this is a minor issue, as finally applied rule ID's will be applied later on cart save.

One more bug I found while fixing this. The bug refers to a Persistent Customer. Here is the test name which helped me with it: StorefrontVerifyShoppingCartPersistenceUnderLongTermCookieTest.

The issue occurs near 
```<!-- 8. Go to Shopping Cart and verify Simple Product 1 is present there  -->```
With my code, we are saving the shopping cart here for the current quote. And now quote address has shipping method. This existing shipping method breaks the overall flow when we click "Not you?" link. Though we have the possibility to save this quote by visiting the checkout. And in vanilla Magento, this issue is also reproducible. I have no access to my Magento right now. In a few days, I can share more details or post a bug.

**Review by rogyar — CHANGES_REQUESTED** [review]
Hi folks. In case if this PR is approved, we need to adjust the MFTF tests according to the best practices. Please, take a look at the following short info https://github.com/magento/magento-functional-tests-migration/wiki/Best-practices and use it as a reference. 

Particularly, I want to pay your attention to the fact that the test should be built using action groups only. We should not use direct actions in a test's body. Following this practice, we create new reusable action groups in the same way as we create reusable service/methods in PHP code. 

Please, take a look at the [following](https://github.com/magento/adobe-stock-integration/blob/bc266993907a33e4b2183558ef53b616a6457b90/AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockListingStateTest.xml#L43) good example of MFTF tests coverage. 

Thank you

**Comment by zhartaunik** [comment]
As I see we are still waiting for architect review?

**Comment by ihor-sviziev** [comment]
> As I see we are still waiting for architect review?

Yes, that's correct https://github.com/magento/magento2/pull/24921/files#r342952196

File: app/code/Magento/Checkout/Model/TotalsInformationManagement.php
Comment (ihor-sviziev) [reply app/code/Magento/Checkout/Model/TotalsInformationManagement.php:74]: Hi @zhartaunik,

It looks like TotalsInformationManagement model in method "calculate" shouldn't save quote, it should execute only calculations. 

Use case: I'd like to _calculate_ how many will I pay for shipment if I'll change region from New York to California --> I'll execute this method, but it doesn't mean I want to save it.

Calculation and re-saving should be still different actions, otherwise it might get performance issue or even worth - in some case infinite recursion.

Could you describe why the way of adding `trigger_recollect` flag was bad and why it didn't worked fine?

**Comment by ihor-sviziev** [comment]
Hi @zhartaunik,

It looks like TotalsInformationManagement model in method "calculate" shouldn't save quote, it should execute only calculations.

Use case: I'd like to calculate how many will I pay for shipment if I'll change region from New York to California --> I'll execute this method, but it doesn't mean I want to save it.

Calculation and re-saving should be still different actions, otherwise it might get performance issue or even worth - in some case infinite recursion.

Could you describe why the way of adding trigger_recollect flag was bad and why it didn't worked fine?

File: app/code/Magento/Checkout/Model/TotalsInformationManagement.php
Comment (slavvka) [inline app/code/Magento/Checkout/Model/TotalsInformationManagement.php:68]: What is the reason of changing this condition?

File: app/code/Magento/Checkout/Model/TotalsInformationManagement.php
Comment (slavvka) [reply app/code/Magento/Checkout/Model/TotalsInformationManagement.php:74]: This method definitely should not save the quote. Probably the best solution is to try to fix it while collecting the totals. @zhartaunik Could dig int that direction?

**Comment by ihor-sviziev** [comment]
I see that this issue already fixed by internal team in 2.3-develop branch (internal) and soon fix will be available in 2.4-develop.

https://github.com/magento/magento2/issues/24526#issuecomment-568109415
https://github.com/magento/magento2/issues/24526#issuecomment-570380866

**Comment by sdzhepa** [comment]
Hello @zhartaunik and all other participants,

Thank you for contribution and collaboration!

Unfortunately, I have to close this PR because of the internal team has already fixed it.
Please see for more details  https://github.com/magento/magento2/issues/24526#issuecomment-575359348

**Pull request closed without merging**
